Blinding/src/MergeDigis_module.cc

```diff
@@ -107,7 +131,168 @@ namespace mu2e{
+      const auto crvDigisMC = event.getValidHandle<CrvDigiMCCollection>(_crv_digimc_tags.at(iTag));
+
+      if(crvDigis->size()!=crvDigisMC->size())
+        throw cet::exception("MergeDigis::mergeCrvDigis") << "mismatched CrvDigi/CrvDigiMC collection sizes" << std::endl;
+
+      for(size_t i=0; i<crvDigis->size(); ++i)
+      {
+        const auto &digi=crvDigis->at(i);
+        const auto &digiMC=crvDigisMC->at(i);
+
+        const CRSScintillatorBarIndex &barIndex = digi.GetScintillatorBarIndex();
+        int SiPM = digi.GetSiPMNumber();
+        size_t channel = barIndex.asUint()*CRVId::nChanPerBar + SiPM;
+
+        digiPairMap[channel].emplace_back(digi,digiMC);
+      }
+    }//loop through collections
+
+//Loop through all channels
+    for(auto digiPairsIter=digiPairMap.begin(); digiPairsIter!=digiPairMap.end(); ++digiPairsIter)
+    {
+      size_t channel  = digiPairsIter->first;
+      std::vector<DigiPair> &digiPairs = digiPairsIter->second;
+
+//Sort digis of a channel by TDC start time and keep digiMCs index-matched
+      std::sort(digiPairs.begin(),digiPairs.end(), [](const DigiPair &d1, const DigiPair &d2) {return d1.first.GetStartTDC()<d2.first.GetStartTDC();});
+
+//Loop through digis of a channel to solve overlap problems that may have occured during merging
+      for(size_t i=1; i<digiPairs.size(); ++i)
+      {
+        const auto &digi1=digiPairs.at(i-1).first;
+        const auto &digi2=digiPairs.at(i).first;
+        const auto &digiMC1=digiPairs.at(i-1).second;
+        const auto &digiMC2=digiPairs.at(i).second;
+        size_t start1=digi1.GetStartTDC();
+        size_t start2=digi2.GetStartTDC();
+        size_t length1=digi1.GetADCs().size();
+        size_t length2=digi2.GetADCs().size();
+        if(start1+length1>start2)
+        {
+//found an overlap between digiPair_1 and digiPair_2
+          double pedestal = calib.pedestal(channel);
+          size_t lengthTotal=std::max(start1+length1,start2+length2)-start1;
+          size_t startOverlap=start2-start1;
+
+//combine ADCs and voltages
+          std::vector<int16_t> ADCs;
+          std::vector<double>  voltages;
+          for(size_t j=0; j<lengthTotal; ++j)
+          {
+            bool hasDigi1 = (j<length1);
+            bool hasDigi2 = (startOverlap<=j && j<startOverlap+length2);
+            if(hasDigi1 && !hasDigi2)
+            {
+              ADCs.push_back(digi1.GetADCs().at(j));
+              voltages.push_back(digiMC1.GetVoltages().at(j));
+            }
+            if(!hasDigi1 && hasDigi2)
+            {
+              ADCs.push_back(digi2.GetADCs().at(j-startOverlap));
+              voltages.push_back(digiMC2.GetVoltages().at(j-startOverlap));
+            }
+            if(hasDigi1 && hasDigi2)
+            {
+              //pedestal is included in both ADC values, but can be used only once. so it needs to be subtracted.
+              int16_t ADC = digi1.GetADCs().at(j) + digi2.GetADCs().at(j-startOverlap) - pedestal;
```

Comment:
Is there any waveform saturation (when the analog signal value exceeds the word size of the adc) that needs to be put here, to "cap off" big pulses after summing?